ir/src/depgraph.rs

```diff
@@ -0,0 +1,433 @@
+                })
+                .collect(),
+        }
+    }
+}
+
+/// SystemTime lacks a platform independent representation we can
+/// depend on so use FileTime's unix_seconds,nanos.
+/// unix_seconds = nanos = size = 0 is used to represent lack of a DepGraphEntry.
+#[derive(Serialize, Deserialize, Debug, Clone)]
+struct DepGraphEntrySerdeRepr {
+    path: String,
+    unix_seconds: i64,
+    nanos: u32,
+    size: u64,
+}
+
+impl DepGraphEntrySerdeRepr {
+    fn has_graph_entry(&self) -> bool {
+        !(self.unix_seconds == 0 && self.nanos == 0 && self.size == 0)
+    }
+}
+
+pub struct Change {
+    path: PathBuf,
+    updated_entry: DepGraphEntry,
+}
+
+pub enum InitialState {
+    NotChanged,
+    Changed,
+}
+
+impl DepGraphEntry {
+    fn new(path: &Path) -> Result<DepGraphEntry, io::Error> {
+        let metadata = path.metadata()?;
+        Ok(DepGraphEntry {
+            mtime: FileTime::from_system_time(metadata.modified()?),
+            size: metadata.len(),
+        })
+    }
+}
+
+impl DepGraph {
+    pub fn new() -> DepGraph {
+        DepGraph {
+            entries: HashMap::new(),
+        }
+    }
+}
+
+impl Default for DepGraph {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DepGraph {
+    pub fn is_tracked(&self, path: &PathBuf) -> bool {
+        self.entries.contains_key(path)
+    }
+
+    pub fn track(&mut self, path: &PathBuf, default_state: InitialState) -> Result<(), io::Error> {
+        if !self.entries.contains_key(path) {
+            self.entries.insert(
+                path.clone(),
+                match default_state {
+                    InitialState::NotChanged => Some(DepGraphEntry::new(path)?),
+                    InitialState::Changed => None,
+                },
+            );
+
+            // For a dir to register unchanged we need to add it's current contents
+            if let InitialState::NotChanged = default_state {
+                if path.is_dir() {
+                    let mut dirs_visited = HashSet::new();
+                    if let Some(new_paths) = self.new_paths(&mut dirs_visited, path) {
+                        for new_path in new_paths {
+                            self.entries
+                                .insert(new_path.clone(), Some(DepGraphEntry::new(&new_path)?));
+                        }
+                    }
+                }
+            }
+        }
+        Ok(())
+    }
+
+    pub fn has_changed(&self, change: &Change) -> bool {
+        return self
+            .entries
+            .get(&change.path)
+            // If we are tracking with different state or we're not tracking at all then you've changed
+            .map(|maybe_entry| {
+                maybe_entry
+                    .as_ref()
+                    .map(|entry| change.updated_entry != *entry)
+                    .unwrap_or(true)
+            })
+            .unwrap_or(true);
+    }
+
+    pub fn update(&mut self, changes: &Vec<Change>) {
```

Comment:
It is a convention to avoid `&Vec` in favour of `&[]`, which is strictly more flexible (unless  you need mutability)

Reply:
TIL, ty.